### PR TITLE
feat: DID document

### PR DESCRIPTION
**Pull request by Dudleyneedham**
## fixes https://github.com/KILTprotocol/ticket/issues/32

Added the DID document to the demo client.

NB: small issue unrelated to this ticket: if we delete the Identity w/ a DID & Document, and add the Identity again the DID & Document is not added. Opened a new Ticket to fix it. Have an idea that it comes from the storing of the contacts and retrieval of the contacts that it isn't picking up the DID etc.
Ticket: KILTprotocol/ticket#273

## How to test:

No test at the moment.

## Checklist:

- [x] I have verified that the code works
- [x] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [x] I have [left the code in a better state](https://deviq.com/boy-scout-rule/)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

File: src/services/ContactRepository.ts [hunk 1]
```diff
@@ -63,8 +62,16 @@ class ContactRepository {
       })
       .then(response => response.json())
       .then((contact: Contact) => {
+        const dID = contact.did
+
         PersistentStore.store.dispatch(Contacts.Store.addContact(contact))
-        return contact
+
+        return {
```
Comment (maudnals) [inline src/services/ContactRepository.ts]: Why is this needed, since `contact` already has a `did` property?

File: src/services/DidService.ts [hunk 2]
```diff
@@ -6,6 +6,7 @@ import BlockchainService from './BlockchainService'
 import ContactRepository from './ContactRepository'
 import MessageRepository from './MessageRepository'
 import { object } from 'prop-types'
+import PersistentStore from '../state/PersistentStore'
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/services/DidService.ts]: No,  I was just testing something. I believe I just forgot to delete it. Thank you.

File: src/services/DidService.ts [hunk 3]
```diff
@@ -36,7 +38,7 @@ export class DidService {
     } as Contact)
 
     const status = await did.store(myIdentity.identity)
-    if (status.type !== 'Finalised') {
+    if (status.type !== 'Finalized') {
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/services/DidService.ts]: Yes, I haven't done anything with it but I will later.

File: src/containers/DidView/DidView.scss [hunk 4]
```diff
@@ -0,0 +1,41 @@
+.DidView {
+  & h1 {
+    @include h1('Wallet');
+  }
+
+  & table {
+    $label-width: 30%;
+
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/containers/DidView/DidView.scss]: I don't think so.

File: src/containers/DidView/DidView.tsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import './DidView.scss'
+import ContactRepository from '../../services/ContactRepository'
+
+type Props = RouteComponentProps<{}> & {
+  selectedIdentity: Wallet.Entry
+}
+
+type State = {
+  contact?: Contact
+  did?: Contact['did'] | null
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/containers/DidView/DidView.tsx]: Ok, I will update this.

File: src/services/ContactRepository.ts (hunk 1 above)
Comment (Dudleyneedham) [reply src/services/ContactRepository.ts]: It doesn't or at least at some point it is lost.

**Comment by Dudleyneedham** [comment]
Screenshot 2020-01-23 at 11 18 06

I have updated the wallet and Identity view to take the did document.

File: src/components/IdentityView/IdentityView.tsx [hunk 6]
```diff
@@ -105,7 +106,11 @@ class IdentityView extends React.Component<Props, State> {
             <label>DID</label>
             <div>
               {myIdentity.did ? (
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/components/IdentityView/IdentityView.tsx]: I agree I will update as suggested

File: src/containers/DidView/DidView.scss (hunk 4 above)
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/containers/DidView/DidView.scss]: I removed the lines manually

File: src/containers/DidView/DidView.scss [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    width: 100%;
+
+    & td {
+      &:first-child {
+        width: $label-width;
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/containers/DidView/DidView.scss]: Not sure if the css is being used now.

File: src/containers/DidView/DidView.scss [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      }
+
+      &:last-child {
+        width: 100% - $label-width;
+      }
+    }
+
+    & tr {
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/containers/DidView/DidView.scss]: No, I have done it manually.

File: src/containers/DidView/DidView.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/containers/DidView/DidView.tsx]: I think we will be speaking with Timo about this.

File: src/containers/DidView/DidView.tsx [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+class DidView extends React.Component<Props, State> {
+  constructor(props: Props) {
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/containers/DidView/DidView.tsx]: I have completely changed this to a Dumb Component. In the end, this was much easier to work with and now I understand that it didn't have any functionality. I was just copying or basing my work from within the Demo-Client itself. It feels that some of these are also needing to only be Dumb components.

File: src/containers/DidView/DidView.tsx [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  public componentDidUpdate(prevProps: Props) {
+    if (!_.isEqual(this.props, prevProps)) {
+      this.setDid()
+    }
+  }
+
+  public render() {
+    const identity = this.props.selectedIdentity
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/containers/DidView/DidView.tsx]: Done.

File: src/containers/DidView/DidView.tsx [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    ContactRepository.findByAddress(selectedIdentity.identity.address).then(
+      (contact: Contact) => {
+        this.setState({ did: contact.did })
+      }
+    )
+  }
+}
+
+const mapStateToProps = (state: ReduxState) => ({
+  identities: Wallet.getAllIdentities(state),
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/containers/DidView/DidView.tsx]: Agreed. I have updated.

File: src/services/ContactRepository.ts [hunk 12]
```diff
@@ -127,6 +132,7 @@ class ContactRepository {
         },
       }
     )
+    // When importing a DID do you think we should be creating the document here as well?
```
Comment (Dudleyneedham) [inline src/services/ContactRepository.ts]: @maudnals Could you please, let me know if I should be creating a DID document here when I am importing? It seems I should be.

Comment (Dudleyneedham) [reply src/services/ContactRepository.ts]: As we import it, but what does it do?

**Comment by Dudleyneedham** [comment]
Screenshot 2020-01-23 at 12 26 15

@tjwelde This is what you want?

File: src/containers/DidView/DidView.tsx [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+import { Contact } from '../../types/Contact'
+import './DidView.scss'
+import Code from '../../components/Code/Code'
+
+type Props = {
+  did: Contact['did']
+  children: any
+}
+
+const DidView = (did: Props) => {
```
Comment (maudnals) [inline src/containers/DidView/DidView.tsx]: Change request:
Please rename this to `DidDocumentView`; what you're displaying really is the DID Doc.

File: src/components/IdentityView/IdentityView.tsx [hunk 14]
```diff
@@ -79,45 +81,49 @@ class IdentityView extends React.Component<Props, State> {
...
-            <div>{myIdentity.identity.signPublicKeyAsHex}</div>
+            <div>{identity.signPublicKeyAsHex}</div>
           </div>
           <div>
             <label>Encryption Public Key</label>
-            <div>{myIdentity.identity.boxPublicKeyAsHex}</div>
+            <div>{identity.boxPublicKeyAsHex}</div>
           </div>
           <div>
             <label>DID</label>
```
Comment (maudnals) [inline src/components/IdentityView/IdentityView.tsx]: Your UI looks good!
However DID and DID Document are not the same thing, so the label is not quite right. 
I'd suggest to append `Document` to the `<label>` when the DID Doc is displayed.

File: src/services/ContactRepository.ts (hunk 12 above)
Comment (maudnals) [reply src/services/ContactRepository.ts]: YES! NO! Let's talk, I'm not sure to understand? 🐑

File: src/components/IdentityView/IdentityView.tsx [hunk 15]
```diff
@@ -79,45 +81,49 @@ class IdentityView extends React.Component<Props, State> {
...
-              {myIdentity.did ? (
-                <span className="did">{myIdentity.did}</span>
+              {did ? (
+                <span className="did">
+                  <DidView did={did.document}>{did.address}</DidView>
+                </span>
               ) : (
-                ''
+                <>
+                  <div>Identity doesn't own a DID.</div>
```
Comment (maudnals) [inline src/components/IdentityView/IdentityView.tsx]: Suggestion: 
No DID is attached to this identity

File: src/services/ContactRepository.ts (hunk 12 above)
Comment (maudnals) [reply src/services/ContactRepository.ts]: Nice talk --> we agree that `ContactRepository` shouldn't be part of this PR. 🦑

File: src/routes/index.tsx [hunk 16]
```diff
@@ -19,6 +19,7 @@ import WalletView from '../containers/WalletView/WalletView'
 import requiresIdentity from './RequiresIdentity'
 import DelegationCreate from '../containers/DelegationCreate/DelegationCreate'
 import Setup from '../containers/Setup/Setup'
+import DidView from '../containers/DidView/DidView'
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/routes/index.tsx]: Removed!

File: src/containers/DidView/DidView.tsx (hunk 13 above)
Comment (Dudleyneedham) [reply src/containers/DidView/DidView.tsx]: I have updated everything to match.

File: src/components/IdentityView/IdentityView.tsx (hunk 14 above)
Comment (Dudleyneedham) [reply src/components/IdentityView/IdentityView.tsx]: I have updated it, so it now looks like `DID` - When there isn't a `DID`. But, `DID Document`.

Comment (Dudleyneedham) [reply src/components/IdentityView/IdentityView.tsx]: Screenshot 2020-01-23 at 14 32 13

File: src/components/IdentityView/IdentityView.tsx (hunk 15 above)
Comment (Dudleyneedham) [reply src/components/IdentityView/IdentityView.tsx]: Updated!

File: src/containers/DidView/DidView.tsx [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+  return (
+    <>
+      <div className="attributes">
+        <div>
+          <Code>{did.did}</Code>{' '}
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/containers/DidView/DidView.tsx]: I have updated it, would like your feedback on it now.

File: src/components/IdentityView/IdentityView.scss [hunk 18]
```diff
@@ -2,11 +2,10 @@
   position: relative;
   padding: $base-padding;
   margin: {
-    bottom: 2*$base-padding;
+    bottom: 2 * $base-padding;
```
Comment (maudnals) [inline src/components/IdentityView/IdentityView.scss]: Question:
Why / Has the prettier config changed?

File: src/state/ducks/Contacts.ts [hunk 19]
```diff
@@ -78,10 +78,8 @@ class Store {
 
         const contact = state.getIn(['contacts', address])
         const { metaData, publicIdentity } = contact
-
```
Comment (maudnals) [inline src/state/ducks/Contacts.ts]: Change request:
In general, if your ticket doesn't affect a file functionally, the file shouldn't be part of your PR. It makes it easier for you and your reviewer!

File: src/services/ContactRepository.ts [hunk 20]
```diff
@@ -53,7 +53,6 @@ class ContactRepository {
     if (persistedContact) {
       return persistedContact
     }
-
```
Comment (maudnals) [inline src/services/ContactRepository.ts]: Change request:
In general, if your ticket doesn't affect a file functionally, the file shouldn't be part of your PR. It makes it easier for you and your reviewer!

File: src/routes/index.tsx [hunk 21]
```diff
@@ -87,7 +87,9 @@ const Routes: React.FunctionComponent<{}> = props => {
         />
         <Route
           path={'/pcrs'}
-          component={requiresIdentity(DelegationsView, { isPCR: true })}
```
Comment (maudnals) [inline src/routes/index.tsx]: Change request:
In general, if your ticket doesn't affect a file functionally, the file shouldn't be part of your PR. It makes it easier for you and your reviewer!

File: src/components/IdentityView/IdentityView.tsx [hunk 22]
```diff
@@ -1,12 +1,14 @@
 import * as React from 'react'
 import { connect } from 'react-redux'
 
+import { Link } from 'react-router-dom'
```
Comment (maudnals) [inline src/components/IdentityView/IdentityView.tsx]: Question:
Why this import?
Make sure your linter is catching these, it saves you time.

File: src/components/IdentityView/IdentityView.tsx [hunk 23]
```diff
@@ -221,7 +233,7 @@ class IdentityView extends React.Component<Props, State> {
 
   private toggleContacts() {
     const { contacts, myIdentity } = this.props
-
+    let title = 'Add Contact Favourites'
```
Comment (maudnals) [inline src/components/IdentityView/IdentityView.tsx]: Question:
Where is this used?

File: src/containers/DidDocumentView/DidDocumentView.tsx [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+type Props = {
+  did: Contact['did']
+  children: any
+}
+
+const DidDocumentView = (props: Props) => {
+  const { did } = props
+  return (
+    <>
+      <div className="attributes">
```
Comment (maudnals) [inline src/containers/DidDocumentView/DidDocumentView.tsx]: Question:
Did you see my comment on the previous review? Looks like you resolved it, but it's unchanged.

File: src/containers/DidDocumentView/DidDocumentView.tsx [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+        <div>
+          <Code>{did?.document}</Code>{' '}
```
Comment (maudnals) [inline src/containers/DidDocumentView/DidDocumentView.tsx]: Change request:
`{' '}` is unnecessary

File: src/components/IdentityView/IdentityView.tsx (hunk 23 above)
Comment (Dudleyneedham) [reply src/components/IdentityView/IdentityView.tsx]: Woops, I was trialling something. Thank you

File: src/routes/index.tsx (hunk 21 above)
Comment (Dudleyneedham) [reply src/routes/index.tsx]: Ok, I think my Pretty is active again. Must have noticed the needed change.

File: src/components/IdentityView/IdentityView.scss (hunk 18 above)
Comment (Dudleyneedham) [reply src/components/IdentityView/IdentityView.scss]: Yes, it did.

File: src/components/IdentityView/IdentityView.tsx (hunk 22 above)
Comment (Dudleyneedham) [reply src/components/IdentityView/IdentityView.tsx]: Ok, it wasn't and still doesn't.

File: src/containers/DidDocumentView/DidDocumentView.tsx (hunk 24 above)
Comment (Dudleyneedham) [reply src/containers/DidDocumentView/DidDocumentView.tsx]: Ok, Iterated again.

File: src/containers/DidDocumentView/DidDocumentView.tsx (hunk 25 above)
Comment (Dudleyneedham) [reply src/containers/DidDocumentView/DidDocumentView.tsx]: Agreed. Lint added it.

File: src/services/ContactRepository.ts (hunk 1 above)
Comment (Dudleyneedham) [reply src/services/ContactRepository.ts]: Yes, I have fixed it.

File: src/state/ducks/Contacts.ts (hunk 19 above)
Comment (Dudleyneedham) [reply src/state/ducks/Contacts.ts]: Yes, I have updated it.

File: src/services/ContactRepository.ts (hunk 20 above)
Comment (Dudleyneedham) [reply src/services/ContactRepository.ts]: Yes, I have updated it.

File: src/routes/index.tsx (hunk 21 above)
Comment (Dudleyneedham) [reply src/routes/index.tsx]: I have updated it.

File: src/services/DidService.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (maudnals) [reply src/services/DidService.ts]: https://github.com/KILTprotocol/ticket/issues/221

File: src/services/DidService.ts [hunk 26]
```diff
@@ -36,23 +35,23 @@ export class DidService {
...
       throw new Error(
         `Error creating DID for identity ${myIdentity.metaData.name}`
       )
     }
 
     persistentStore.store.dispatch(
       Wallet.Store.updateIdentityAction(myIdentity.identity.address, {
-        did: did.identifier,
+        did: { address: did.identifier, document: didDocument },
```
Comment (maudnals) [inline src/services/DidService.ts]: Change request:
identifier (`kilt:did:abcd`) !== address (`abcd`), let's make this clear!

File: src/containers/DidDocumentView/DidDocumentView.tsx [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+
+type Props = {
+  didDocument: object
+  children: any
+}
+
+const DidDocumentView = (props: Props) => {
+  const { didDocument } = props
+  return (
+    <>
```
Comment (maudnals) [inline src/containers/DidDocumentView/DidDocumentView.tsx]: Question:
Do you really need to wrap this in a fragment? Why not return `<div className="attributes">` directly?

File: src/components/IdentityView/IdentityView.scss (hunk 18 above)
Comment (maudnals) [reply src/components/IdentityView/IdentityView.scss]: Question:
As just discussed live, can you please check what's off with this?

File: src/containers/DidDocumentView/DidDocumentView.tsx (hunk 27 above)
Comment (Dudleyneedham) [reply src/containers/DidDocumentView/DidDocumentView.tsx]: Yes, the fragment does nothing. I have removed it.

File: src/services/DidService.ts (hunk 26 above)
Comment (Dudleyneedham) [reply src/services/DidService.ts]: I have changed this to Identifier everywhere. It was this before. 

What do you think @maudnals ?

File: src/services/ContactRepository.ts [hunk 28]
```diff
@@ -133,7 +133,7 @@ class ContactRepository {
         PersistentStore.store.getState()
       )
       const contact = {
-        did: { address: identifier },
+        did: { identifier: identifier },
```
Comment (maudnals) [inline src/services/ContactRepository.ts]: Change request: 
Change syntax to "identifier" directly

File: src/containers/DidDocumentView/DidDocumentView.tsx [hunk 29]
```diff
@@ -0,0 +1,22 @@
+import * as React from 'react'
+import { Contact } from '../../types/Contact'
+import './DidDocumentView.scss'
+import Code from '../../components/Code/Code'
+
+type Props = {
+  didDocument: object
+  children: any
```
Comment (maudnals) [inline src/containers/DidDocumentView/DidDocumentView.tsx]: Question:
I know you're using the children, but do you need to explicitly list them as prop?

File: src/containers/DidDocumentView/DidDocumentView.tsx [hunk 30]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+const DidDocumentView = (props: Props) => {
+  const { didDocument } = props
+  return (
+    <div className="attributes">
+      <div>
```
Comment (maudnals) [inline src/containers/DidDocumentView/DidDocumentView.tsx]: Question:
Do you need this div?

File: src/services/DidService.ts (hunk 26 above)
Comment (maudnals) [reply src/services/DidService.ts]: Great! Please make sure it doesn't break anything else

File: src/containers/DidDocumentView/DidDocumentView.tsx (hunk 30 above)
Comment (Dudleyneedham) [reply src/containers/DidDocumentView/DidDocumentView.tsx]: No, it isn't needed. It has been removed.

File: src/containers/DidDocumentView/DidDocumentView.tsx (hunk 29 above)
Comment (Dudleyneedham) [reply src/containers/DidDocumentView/DidDocumentView.tsx]: Yes, if I remove it. I receive an error.

`"Type '{ children: string | undefined; didDocument: object; }' is not assignable to type 'IntrinsicAttributes & Props'.\n  Property 'children' does not exist on type 'IntrinsicAttributes & Props'."`

Tried to understand why it happens, but I wasn't able too. However, adding the children etc works.

File: src/services/ContactRepository.ts (hunk 28 above)
Comment (Dudleyneedham) [reply src/services/ContactRepository.ts]: Done.

File: src/containers/DidDocumentView/DidDocumentView.tsx (hunk 29 above)
Comment (maudnals) [reply src/containers/DidDocumentView/DidDocumentView.tsx]: OK got it, good that you checked; it's because you're using the children, so TypeScript needs to explicitely know about it.

**Comment by Dudleyneedham** [comment]
Noticed that if we delete the Identity w/ a DID & Document, and add the Identity again the DID & Document is not added. Opened a new Ticket to fix it.

Have an idea that it comes from the storing of the contacts and retrieval of the contacts that it isn't picking up the DID etc.
